File: leetcode/easy/0009-palindrome-number/solution.kt
class Solution {
    fun isPalindrome(x: Int): Boolean {
        // Negative numbers cannot be palindrome
        if (x < 0) return false
        
        // Numbers ending with 0 (except 0 itself) cannot be palindrome
        if (x % 10 == 0 && x != 0) return false
        
        var num = x
        var reversed = 0
        
        // Reverse only half of the number
        while (num > reversed) {
            reversed = reversed * 10 + num % 10
            num /= 10
        }
        
        // For even digits: num == reversed
        // For odd digits: num == reversed / 10 (middle digit ignored)
        return num == reversed || num == reversed / 10
    }
}

File: leetcode/easy/0009-palindrome-number/solution.cpp
class Solution {
public:
    bool isPalindrome(int x) {
        // Negative numbers are not palindrome
        if (x < 0) return false;

        // Numbers ending with 0 (except 0) are not palindrome
        if (x % 10 == 0 && x != 0) return false;

        int reversed = 0;
        int original = x;

        while (x > 0) {
            int digit = x % 10;
            reversed = reversed * 10 + digit;
            x /= 10;
        }

        return original == reversed;
    }
};